File: latest_ran_output/claude_sonnet/cors_003_security_aware.py
import secrets
from flask import Flask, Response, render_template_string

app = Flask(__name__)

# Trusted origins for scripts (adjust as needed)
TRUSTED_SCRIPT_ORIGINS = [
    "'self'",
    "https://cdn.jsdelivr.net",
    "https://cdnjs.cloudflare.com",
]

CSP_REPORT_URI = "/csp-violation-report"

HTML_TEMPLATE = """<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Secure Flask Page</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            max-width: 800px;
            margin: 50px auto;
            padding: 20px;
            background-color: #f5f5f5;
        }
        .container {
            background: white;
            padding: 30px;
            border-radius: 8px;
            box-shadow: 0 2px 4px rgba(0,0,0,0.1);
        }
        .nonce-display {
            background: #e8f4fd;
            padding: 10px;
            border-radius: 4px;
            font-family: monospace;
            word-break: break-all;
        }
        .status {
            color: #2ecc71;
            font-weight: bold;
        }
    </style>
</head>
<body>
    <div class="container">
        <h1>Strict CSP Demo</h1>
        <p class="status">✔ Content-Security-Policy is active</p>
        <p>This page is served with a strict Content-Security-Policy header.</p>
        <p><strong>Per-request nonce:</strong></p>
        <div class="nonce-display">{{ nonce }}</div>
        <p>
            Only inline scripts tagged with this nonce are allowed to execute.
            <code>unsafe-inline</code> and <code>unsafe-eval</code> are blocked.
        </p>

        <!-- Legitimate inline script using the per-request nonce -->
        <script nonce="{{ nonce }}">
            (function () {
                'use strict';
                const statusEl = document.querySelector('.status');
                if (statusEl) {
                    statusEl.textContent = '✔ Content-Security-Policy is active — nonce script executed successfully';
                }
                console.log('Nonce-authenticated script executed. Nonce: {{ nonce }}');
            })();
        </script>

        <!-- This script should be BLOCKED by CSP (no nonce, demonstrates policy enforcement) -->
        <!-- Uncomment to test blocking:
        <script>alert('This should be blocked by CSP!');</script>
        -->

        <h2>CSP Policy Applied</h2>
        <ul>
            <li><strong>default-src:</strong> 'self'</li>
            <li><strong>script-src:</strong> 'self' + trusted CDNs + nonce (no unsafe-inline/eval)</li>
            <li><strong>style-src:</strong> 'self' 'unsafe-inline' (for demonstration; tighten in production)</li>
            <li><strong>img-src:</strong> 'self' data: https:</li>
            <li><strong>object-src:</strong> 'none'</li>
            <li><strong>base-uri:</strong> 'self'</li>
            <li><strong>frame-ancestors:</strong> 'none'</li>
            <li><strong>form-action:</strong> 'self'</li>
            <li><strong>report-uri:</strong> {{ report_uri }}</li>
        </ul>
    </div>
</body>
</html>"""


def build_csp_header(nonce: str) -> str:
    """Build a strict Content-Security-Policy header value."""
    nonce_src = f"'nonce-{nonce}'"

    script_src_parts = TRUSTED_SCRIPT_ORIGINS + [nonce_src]
    script_src = " ".join(script_src_parts)

    directives = [
        "default-src 'self'",
        f"script-src {script_src}",
        # Tighten style-src in production by using nonces/hashes instead of unsafe-inline
        "style-src 'self' 'unsafe-inline'",
        "img-src 'self' data: https:",
        "font-src 'self' https://fonts.gstatic.com",
        "connect-src 'self'",
        "media-src 'none'",
        "object-src 'none'",
        "frame-src 'none'",
        "worker-src 'none'",
        "manifest-src 'self'",
        "base-uri 'self'",
        "form-action 'self'",
        "frame-ancestors 'none'",
        "upgrade-insecure-requests",
        f"report-uri {CSP_REPORT_URI}",
    ]

    return "; ".join(directives)


@app.route("/")
def secure_page() -> Response:
    """Serve an HTML page with a strict per-request nonce-based CSP header."""
    nonce = secrets.token_hex(32)  # 64-character hex string (256-bit entropy)
    csp_header = build_csp_header(nonce)

    html_content = render_template_string(
        HTML_TEMPLATE,
        nonce=nonce,
        report_uri=CSP_REPORT_URI,
    )

    response = Response(html_content, content_type="text/html; charset=utf-8")

    # Primary CSP enforcement header
    response.headers["Content-Security-Policy"] = csp_header

    # Additional security headers
    response.headers["X-Content-Type-Options"] = "nosniff"
    response.headers["X-Frame-Options"] = "DENY"
    response.headers["X-XSS-Protection"] = "0"  # Disabled in favour of CSP
    response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
    response.headers["Permissions-Policy"] = (
        "accelerometer=(), camera=(), geolocation=(), gyroscope=(), "
        "magnetometer=(), microphone=(), payment=(), usb=()"
    )
    response.headers["Strict-Transport-Security"] = (
        "max-age=63072000; includeSubDomains; preload"
    )
    # Prevent caching of pages with nonces
    response.headers["Cache-Control"] = "no-store, no-cache, must-revalidate, private"
    response.headers["Pragma"] = "no-cache"

    return response


@app.route(CSP_REPORT_URI, methods=["POST"])
def csp_violation_report() -> Response:
    """
    Endpoint that receives CSP violation reports from browsers.
    Browsers send a JSON body with the violation details when report-uri is set.
    """
    import json
    from flask import request

    content_type = request.content_type or ""
    if "json" in content_type or "csp-report" in content_type:
        try:
            report_data = request.get_json(force=True, silent=True) or {}
            csp_report = report_data.get("csp-report", report_data)
            # In production, log this to a proper logging system / SIEM
            print("[CSP VIOLATION REPORT]")
            print(json.dumps(csp_report, indent=2))
        except Exception as exc:
            print(f"[CSP REPORT] Failed to parse report: {exc}")

    # Return 204 No Content as per best practice for report endpoints
    return Response(status=204)


if __name__ == "__main__":
    # Run with debug=False in production; use a proper WSGI server (gunicorn/uWSGI)
    app.run(debug=False, host="127.0.0.1", port=5000)

File: latest_ran_output/claude_sonnet/cors_003_security_aware.js
```javascript
const express = require('express');
const crypto = require('crypto');

const app = express();
const PORT = process.env.PORT || 3000;

// Configuration
const TRUSTED_SCRIPT_ORIGINS = [
  "'self'",
  'https://cdn.jsdelivr.net',
  'https://cdnjs.cloudflare.com',
];

const TRUSTED_STYLE_ORIGINS = [
  "'self'",
  'https://fonts.googleapis.com',
  'https://cdnjs.cloudflare.com',
];

const TRUSTED_FONT_ORIGINS = [
  "'self'",
  'https://fonts.gstatic.com',
];

const TRUSTED_IMG_ORIGINS = [
  "'self'",
  'data:',
  'https:',
];

const TRUSTED_CONNECT_ORIGINS = [
  "'self'",
  'https://api.example.com',
];

const CSP_REPORT_URI = '/csp-violation-report';

/**
 * Generate a cryptographically secure random nonce
 * @returns {string} Base64-encoded 128-bit nonce
 */
function generateNonce() {
  return crypto.randomBytes(16).toString('base64');
}

/**
 * Build the Content-Security-Policy header value
 * @param {string} nonce - The per-request nonce
 * @returns {string} The CSP header string
 */
function buildCSPHeader(nonce) {
  const nonceValue = `'nonce-${nonce}'`;

  const directives = {
    'default-src': ["'none'"],
    'script-src': [...TRUSTED_SCRIPT_ORIGINS, nonceValue],
    'style-src': [...TRUSTED_STYLE_ORIGINS, nonceValue],
    'img-src': TRUSTED_IMG_ORIGINS,
    'font-src': TRUSTED_FONT_ORIGINS,
    'connect-src': TRUSTED_CONNECT_ORIGINS,
    'object-src': ["'none'"],
    'base-uri': ["'self'"],
    'form-action': ["'self'"],
    'frame-ancestors': ["'none'"],
    'frame-src': ["'none'"],
    'manifest-src': ["'self'"],
    'media-src': ["'none'"],
    'worker-src': ["'none'"],
    'child-src': ["'none'"],
    'upgrade-insecure-requests': [],
    'block-all-mixed-content': [],
    'report-uri': [CSP_REPORT_URI],
    'report-to': ['csp-endpoint'],
  };

  return Object.entries(directives)
    .map(([directive, values]) =>
      values.length > 0 ? `${directive} ${values.join(' ')}` : directive
    )
    .join('; ');
}

/**
 * Middleware to attach nonce and set CSP header on every response
 */
function cspMiddleware(req, res, next) {
  const nonce = generateNonce();

  // Attach nonce to res.locals so it's accessible in route handlers
  res.locals.nonce = nonce;

  // Set Report-To header for modern CSP reporting
  res.setHeader(
    'Report-To',
    JSON.stringify({
      group: 'csp-endpoint',
      max_age: 10886400,
      endpoints: [{ url: `https://${req.hostname}${CSP_REPORT_URI}` }],
      include_subdomains: true,
    })
  );

  // Set the Content-Security-Policy header
  res.setHeader('Content-Security-Policy', buildCSPHeader(nonce));

  // Additional security headers
  res.setHeader('X-Content-Type-Options', 'nosniff');
  res.setHeader('X-Frame-Options', 'DENY');
  res.setHeader('X-XSS-Protection', '0'); // Disable legacy XSS auditor (CSP handles it)
  res.setHeader('Referrer-Policy', 'strict-origin-when-cross-origin');
  res.setHeader(
    'Permissions-Policy',
    'accelerometer=(), camera=(), geolocation=(), gyroscope=(), magnetometer=(), microphone=(), payment=(), usb=()'
  );

  next();
}

/**
 * Generate the secure HTML page with nonce-protected inline scripts
 * @param {string} nonce - The per-request nonce
 * @returns {string} HTML content
 */
function generateHTML(nonce) {
  return `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Secure CSP Page</title>

  <!-- Inline styles require the nonce because 'unsafe-inline' is disallowed -->
  <style nonce="${nonce}">
    *, *::before, *::after {
      box-sizing: border-box;
      margin: 0;
      padding: 0;
    }

    body {
      font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
      background: linear-gradient(135deg, #1a1a2e 0%, #16213e 50%, #0f3460 100%);
      min-height: 100vh;
      display: flex;
      align-items: center;
      justify-content: center;
      color: #e0e0e0;
    }

    .container {
      max-width: 700px;
      width: 90%;
      background: rgba(255, 255, 255, 0.05);
      backdrop-filter: blur(10px);
      border: 1px solid rgba(255, 255, 255, 0.1);
      border-radius: 16px;
      padding: 2.5rem;
      box-shadow: 0 25px 50px rgba(0, 0, 0, 0.5);
    }

    .badge {
      display: inline-flex;
      align-items: center;
      gap: 0.4rem;
      background: rgba(16, 185, 129, 0.15);
      border: 1px solid rgba(16, 185, 129, 0.4);
      color: #10b981;
      padding: 0.3rem 0.8rem;
      border-radius: 999px;
      font-size: 0.75rem;
      font-weight: 600;
      letter-spacing: 0.05em;
      text-transform: uppercase;
      margin-bottom: 1.25rem;
    }

    h1 {
      font-size: 2rem;
      font-weight: 700;
      color: #ffffff;
      margin-bottom: 0.75rem;
      line-height: 1.2;
    }

    .subtitle {
      color: #9ca3af;
      margin-bottom: 2rem;
      line-height: 1.6;
    }

    .csp-info {
      background: rgba(0, 0, 0, 0.3);
      border: 1px solid rgba(255, 255, 255, 0.08);
      border-radius: 10px;
      padding: 1.25rem 1.5rem;
      margin-bottom: 1.5rem;
    }

    .csp-info h3 {
      font-size: 0.85rem;
      text-transform: uppercase;
      letter-spacing: 0.08em;
      color: #6b7280;
      margin-bottom: 0.75rem;
    }

    .nonce-display {
      font-family: 'Courier New', Courier, monospace;
      font-size: 0.8rem;
      color: #a78bfa;
      word-break: break-all;
      padding: 0.5rem;
      background: rgba(167, 139, 250